filebeat/reader/line/line.go

```diff
@@ -52,143 +64,81 @@ func New(input io.Reader, codec encoding.Encoding, bufferSize int) (*Reader, err
 		return nil, err
 	}
 
-	return &Reader{
-		reader:     input,
-		codec:      codec,
+	return &lineReader{
+		reader:     reader,
 		bufferSize: bufferSize,
 		nl:         nl,
-		decoder:    codec.NewDecoder(),
-		inBuffer:   streambuf.New(nil),
-		outBuffer:  streambuf.New(nil),
+		buffer:     streambuf.New(nil),
+		offset:     0,
 	}, nil
 }
 
-// Next reads the next line until the new line character
-func (r *Reader) Next() ([]byte, int, error) {
-	// This loop is need in case advance detects an line ending which turns out
-	// not to be one when decoded. If that is the case, reading continues.
-	for {
-		// read next 'potential' line from input buffer/reader
-		err := r.advance()
-		if err != nil {
-			return nil, 0, err
-		}
+func (l *lineReader) Read(buf []byte) (int, error) {
+	idx := l.buffer.Index(l.nl)
```

Comment:
This `Index` searches by byte sequence. The risk of this is, that the returned index is no actual newline (e.g. if `len(nl) == 1`), but might point in the middle of an encoded symbol. The old code did try to decode up to the `idx` and double check we actually hit a newline symbol.